cmd/server/main.go

```diff
@@ -1,3 +1,170 @@
 package main
 
-func main() {}
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"slices"
+	"strconv"
+
+	"github.com/go-chi/chi/v5"
+)
+
+type gauge float64
+type counter int64
+
+type MemStorage struct {
+	Gauges   map[string]gauge
+	Counters map[string]counter
+}
+
+func (m *MemStorage) SetValue(t, name, value string) error {
+	switch t {
+	case "gauge":
+		i, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return err
+		}
+		m.Gauges[name] = gauge(i)
+	case "counter":
+		i, err := strconv.Atoi(value)
+		if err != nil {
+			return err
+		}
+		v, ok := m.Counters[name]
+		if ok {
+			m.Counters[name] = v + counter(i)
+		} else {
+			m.Counters[name] = counter(i)
+		}
+
+	default:
+		return errors.New("not correct type")
+	}
+	return nil
+}
+
+func (m *MemStorage) GetValue(t, name string) (string, error) {
+	value := ""
+	var err error
+	switch t {
+	case "gauge":
+		v, ok := m.Gauges[name]
+		if ok {
+			value = fmt.Sprint(v)
+		}
+
+	case "counter":
+		v, ok := m.Counters[name]
+		if ok {
+			value = fmt.Sprint(v)
+		}
+	}
+	if value == "" {
+		err = errors.New("not found metric")
+	}
+	return value, err
+}
+
+func (m *MemStorage) ToStringValues() string {
+	result := ""
+
+	for k, v := range m.Gauges {
+		result += fmt.Sprintf("%s: %s\n", k, fmt.Sprint(v))
+	}
+
+	for k, v := range m.Counters {
+		result += fmt.Sprintf("%s: %s\n", k, fmt.Sprint(v))
+	}
+
+	return result
+}
+
+var availableMetricTypes = []string{
+	"gauge",
+	"counter",
+}
+
+func main() {
+	parseFlags()
+	if err := run(); err != nil {
+		panic(err)
+	}
+}
+
+func UpdateMetrics(ms *MemStorage) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodPost {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		mType := chi.URLParam(r, "metric_type")
+		mName := chi.URLParam(r, "metric_name")
+		mValue := chi.URLParam(r, "metric_value")
+
+		if !slices.Contains(availableMetricTypes, mType) {
+			w.WriteHeader(http.StatusBadRequest)
```

Comment:
Можно использовать `http.Error(...)` для возврата сообщения об ошибке.

```suggestion
			http.Error(w, err.Error(), http.StatusBadRequest)
```